### PR TITLE
Consolidate Helix shards for quick test assemblies via MethodLimitMultiplier

### PR DESCRIPTION
**Pull request by MichaelSimons**
## Summary

Adds a `MethodLimitMultiplier` metadata option for `SDKCustomXUnitProject` items to consolidate Helix test shards for assemblies that produce many very short work items. This reduces per-item overhead (payload download, agent allocation, setup, teardown) and makes us better citizens of the shared Helix queues.

## How it works

`AssemblyScheduler` partitions each test assembly into shards based on a `methodLimit` threshold (default 16, or 32 for FullFramework). `MethodLimitMultiplier` simply multiplies that threshold, producing fewer, larger shards. It is set via MSBuild item metadata in `UnitTests.proj` — no new config files.

## Changes

- **`SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs`** — Reads optional `MethodLimitMultiplier` metadata; warns on invalid values
- **`UnitTests.proj`** — Applies multipliers to three assembly groups:
  - **NetAnalyzers** (x5)
  - **TemplateEngine / TemplateSearch** (x20)
  - **Api** — ApiCompatibility, ApiDiff, ApiSymbolExtensions, GenAPI (x10)

## Results

Comparison of latest PR validation run ([build 1407886](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1407886)) against main ([build 1405738](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1405738)).

### Linux x64

| Group | Main Shards (Avg) | PR Shards (Avg / Max) | Saved |
|-------|:-:|:-:|:-:|
| NetAnalyzers | 197 (0m41s) | 70 (1m09s / 3m21s) | 127 |
| TemplateEngine¹ | 50 (0m19s) | 15 (0m33s / 1m23s) | 35 |
| Api¹ | 23 (0m17s) | 5 (0m43s / 1m02s) | 18 |
| **Total** | **515** | **335** | **180** |

### Windows x64

| Group | Main Shards (Avg) | PR Shards (Avg / Max) | Saved |
|-------|:-:|:-:|:-:|
| NetAnalyzers | 197 (1m27s) | 70 (2m36s / 4m57s) | 127 |
| TemplateEngine¹ | 50 (0m45s) | 15 (1m12s / 2m33s) | 35 |
| Api¹ | 23 (0m48s) | 5 (2m28s / 2m42s) | 18 |
| **Total** | **518** | **338** | **180** |

### macOS x64

| Group | Main Shards (Avg) | PR Shards (Avg / Max) | Saved |
|-------|:-:|:-:|:-:|
| NetAnalyzers | 197 (1m09s) | 70 (1m28s / 3m02s) | 127 |
| TemplateEngine¹ | 50 (0m42s) | 15 (0m52s / 1m28s) | 35 |
| Api¹ | 23 (0m36s) | 5 (1m06s / 2m01s) | 18 |
| **Total** | **515** | **335** | **180** |

### FullFramework Windows

| Group | Main Shards (Avg) | PR Shards (Avg / Max) | Saved |
|-------|:-:|:-:|:-:|
| NetAnalyzers | 129 (2m04s) | 40 (3m50s / 5m03s) | 89 |
| TemplateEngine¹ | 33 (0m50s) | 15 (2m15s / 2m46s) | 18 |
| Api¹ | 21 (1m21s) | 9 (2m17s / 4m20s) | 12 |
| **Total** | **348** | **229** | **119** |

### Overall

| Leg | Main Shards | PR Shards | Saved |
|-----|:-:|:-:|:-:|
| Linux x64 | 515 | 335 | **180** |
| Windows x64 | 518 | 338 | **180** |
| macOS x64 | 515 | 335 | **180** |
| FullFx Windows | 348 | 229 | **119** |
| **Total** | **1896** | **1237** | **659** |

Each eliminated work item avoids the fixed per-item overhead of agent allocation, payload download, environment setup, and teardown. While this overhead is non-trivial, it varies by queue and machine and is difficult to measure precisely from a single comparison.

¹ TemplateEngine and Api assemblies are now at one shard per assembly — the minimum possible without cross-assembly consolidation. Further shard reduction for these groups would require changes to the Helix sharding model itself.

### CODE REVIEW

File: test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs [hunk 1]
```diff
@@ -153,7 +153,19 @@ private async Task ExecuteAsync()
             }
 
             var isFullMSBuild = string.Equals(Environment.GetEnvironmentVariable("TestFullMSBuild"), "true", StringComparison.OrdinalIgnoreCase);
-            var scheduler = new AssemblyScheduler(methodLimit: isFullMSBuild ? 32 : 16);
+            var baseMethodLimit = isFullMSBuild ? 32 : 16;
+            if (xunitProject.TryGetMetadata("MethodLimitMultiplier", out string multiplierStr))
+            {
```
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs:160]: Not fixing for code simplicity over better UX for edge cases.

File: test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs [hunk 2]
```diff
@@ -153,7 +153,19 @@ private async Task ExecuteAsync()
...
+                if (int.TryParse(multiplierStr, out int multiplier) && multiplier > 0)
+                {
+                    baseMethodLimit *= multiplier;
```
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs:163]: Not fixing for code simplicity over better UX for edge cases.

File: test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs [hunk 3]
```diff
@@ -153,7 +153,19 @@ private async Task ExecuteAsync()
             }
 
             var isFullMSBuild = string.Equals(Environment.GetEnvironmentVariable("TestFullMSBuild"), "true", StringComparison.OrdinalIgnoreCase);
-            var scheduler = new AssemblyScheduler(methodLimit: isFullMSBuild ? 32 : 16);
+            var baseMethodLimit = isFullMSBuild ? 32 : 16;
+            if (xunitProject.TryGetMetadata("MethodLimitMultiplier", out string multiplierStr))
```
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs:159]: @copilot apply changes based on [this feedback](https://github.com/dotnet/sdk/pull/54115#discussion_r3184173530)

[the remark this answers is not in the document]
Comment (marcpopMSFT) [reply test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs:159]: @MichaelSimons it doesn't look like copilot made the change in your fork. Not sure why.

**Review by marcpopMSFT — APPROVED** [review]

File: test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply test/HelixTasks/SDKCustomCreateXUnitWorkItemsWithTestExclusion.cs:159]: Had copilot cli address the comment. Fixed with https://github.com/dotnet/sdk/pull/54115/changes/799a7c18deb263b3d8ebd3a4c4de1a6eb1996111

**Comment by MichaelSimons** [comment]
/ba-g known intermitent failure in netanalyzer tests
